retrieval/discovery/serverset.go

```diff
@@ -327,12 +329,13 @@ func (tc *zookeeperTreeCache) recursiveNodeUpdate(path string, node *zookeeperTr
 				children: map[string]*zookeeperTreeCacheNode{},
 				done:     make(chan struct{}, 1),
 			}
-		}
-		err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
-		if err != nil {
-			return err
+			err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
```

Comment:
I created a crude test for deleting and recreating a tree in
https://github.com/tommyulfsparre/prometheus/commit/6f28826049018162352e05f8cdbb7a0fa3d6be47. It failed before and is fixed with 136c738. 
At least during testing I received the watch event in the order they where triggered.